app/src/server/api/routers/loggedCalls.router.ts

```diff
@@ -180,4 +185,98 @@ export const loggedCallsRouter = createTRPCRouter({
 
       return tags.map((tag) => tag.name);
     }),
+  export: protectedProcedure
+    .input(
+      z.object({
+        projectId: z.string(),
+        selectedLogIds: z.string().array(),
+        testingSplit: z.number(),
+        selectedExportFormat: z.string(),
+        removeDuplicates: z.boolean(),
+      }),
+    )
+    .mutation(async ({ input, ctx }) => {
+      await requireCanViewProject(input.projectId, ctx);
+
+      // Fetch the real data using Prisma
+      const loggedCallsFromDb = await ctx.prisma.loggedCallModelResponse.findMany({
+        where: {
+          originalLoggedCall: {
+            projectId: input.projectId,
+            id: { in: input.selectedLogIds },
+          },
+        },
+      });
+
+      // Convert the database data into the desired format
+      let formattedLoggedCalls: { input: JsonValue[]; output: JsonValue }[] = loggedCallsFromDb.map(
+        (call) => ({
+          input: (call.reqPayload as unknown as Record<string, unknown>).messages as JsonValue[],
+          output: (call.respPayload as unknown as { choices: { message: unknown }[] }).choices[0]
+            ?.message as JsonValue,
+        }),
+      );
+
+      if (input.removeDuplicates) {
+        const deduplicatedLoggedCalls = [];
+        const loggedCallHashSet = new Set<string>();
+        for (const loggedCall of formattedLoggedCalls) {
+          const loggedCallHash = hashObject(loggedCall);
+          if (!loggedCallHashSet.has(loggedCallHash)) {
+            loggedCallHashSet.add(loggedCallHash);
+            deduplicatedLoggedCalls.push(loggedCall);
+          }
+        }
+        formattedLoggedCalls = deduplicatedLoggedCalls;
+      }
+
+      // Remove duplicate messages from input
+      const inputMessageHashMap = new Map<string, number>();
+      for (const loggedCall of formattedLoggedCalls) {
+        for (const message of loggedCall.input) {
+          const hash = hashObject(message);
+          if (inputMessageHashMap.has(hash)) {
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            inputMessageHashMap.set(hash, inputMessageHashMap.get(hash)! + 1);
+          } else {
+            inputMessageHashMap.set(hash, 0);
+          }
+        }
+      }
+      for (const loggedCall of formattedLoggedCalls) {
+        loggedCall.input = loggedCall.input.filter((message) => {
+          const hash = hashObject(message);
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          return inputMessageHashMap.get(hash)! < formattedLoggedCalls.length;
```

Comment:
can you make a comment that if a message is in one input a bunch of times it could trigger removing it for all of them, even if it isn't present in all of them? edge case but technically incorrect behavior.

Reply:
Good idea